br/pkg/lightning/backend/external/reader.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"time"
+
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/br/pkg/membuf"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/pkg/metrics"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+func readAllData(
+	ctx context.Context,
+	storage storage.ExternalStorage,
+	dataFiles, statsFiles []string,
+	startKey, endKey []byte,
+	bufPool *membuf.Pool,
+	output *memKVsAndBuffers,
+) (err error) {
+	task := log.BeginTask(logutil.Logger(ctx), "read all data")
+	task.Info("arguments",
+		zap.Int("data-file-count", len(dataFiles)),
+		zap.Int("stat-file-count", len(statsFiles)),
+		zap.Binary("start-key", startKey),
+		zap.Binary("end-key", endKey),
+	)
+	defer func() {
+		task.End(zap.ErrorLevel, err)
+	}()
+
+	concurrences, startOffsets, err := getFilesReadConcurrency(
+		ctx,
+		storage,
+		statsFiles,
+		startKey,
+		endKey,
+	)
+	// TODO(lance6716): refine adjust concurrency
+	for i, c := range concurrences {
+		if c < readAllDataConcThreshold {
+			concurrences[i] = 1
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+	var eg errgroup.Group
+	for i := range dataFiles {
+		i := i
+		eg.Go(func() error {
+			return readOneFile(
+				ctx,
+				storage,
+				dataFiles[i],
+				startKey,
+				endKey,
+				startOffsets[i],
+				concurrences[i],
+				bufPool,
+				output,
+			)
+		})
+	}
+	return eg.Wait()
+}
+
+// readOneFile will acquire memory from given bufPool for:
```

Comment:
sorry I forget to delete these comments in previous PR, now they are outdated. Please delete the comment of this function.